zkevm-circuits/src/circuit_tools.rs

```diff
@@ -0,0 +1,1015 @@
+    }
+}
+
+/// Trait around RLC
+pub trait RLCChainable<F> {
+    /// Returns the RLC of itself with a starting rlc/multiplier
+    fn rlc_chain(&self, other: Expression<F>) -> Expression<F>;
+}
+
+impl<F: FieldExt> RLCChainable<F> for (Expression<F>, Expression<F>) {
+    fn rlc_chain(&self, other: Expression<F>) -> Expression<F> {
+        self.0.expr() + self.1.expr() * other.expr()
+    }
+}
+
+/// require_parser
+#[macro_export]
+macro_rules! require_parser {
+    {
+        $cb:expr,
+        lhs = ($($lhs:tt)*)
+        rest = (== $($rhs:tt)*)
+    } => {
+        let description = $crate::concat_with_preamble!(
+            stringify!($($lhs)*),
+            " == ",
+            stringify!($($rhs)*)
+        );
+        $crate::_require!($cb, description, $($lhs)* => $($rhs)*)
+    };
+
+    {
+        $cb:expr,
+        lhs = ($($lhs:tt)*)
+        rest = ($next:tt $($rest:tt)*)
+    } => {
+        $crate::require_parser! {
+            $cb,
+            lhs = ($($lhs)* $next)
+            rest = ($($rest)*)
+        }
+    };
+}
+
+/// _require2
+#[macro_export]
+macro_rules! _require2 {
+    ($cb:expr, $($rest:tt)*) => {{
+        $crate::require_parser! {
+            $cb,
+            lhs = ()
+            rest = ($($rest)*)
+        }
+    }};
+}
+
+/// Creates a dummy constraint builder that cannot be used to add constraints.
+#[macro_export]
+macro_rules! _cb {
+    () => {{
+        ConstraintBuilder::<F>::new(0)
+    }};
+}
+
+/// Concats arguments with preamble consisting of the originating file and line.
+#[macro_export]
+macro_rules! concat_with_preamble {
+    ($($args:expr),* $(,)?) => {{
+        concat!(
+            file!(),
+            ":",
+            line!(),
+            ": ",
+            $(
+                $args,
+            )*
+        )
+    }};
+}
+
+/// Can be used to mark a specific branch as unreachable
+#[macro_export]
+macro_rules! _unreachablex {
+    ($cb:expr $(,$descr:expr)?) => {{
+        let descr = concat_with_preamble!(
+            "unreachable executed",
+            $(
+                ": ",
+                $descr,
+            )*
+        );
+        _require!($cb, descr, true => false)
+    }};
+}
+
+/// _require
+#[macro_export]
+macro_rules! _require {
+    ($cb:expr, $lhs:expr => bool) => {{
+        $cb.require_boolean(
+            concat_with_preamble!(
+                stringify!($lhs),
+                " => ",
+                "bool",
+            ),
+            $lhs.expr(),
+        );
+    }};
+
+    ($cb:expr, $lhs:expr => $rhs:expr) => {{
+        let description = concat_with_preamble!(
+            stringify!($lhs),
+            " => ",
+            stringify!($rhs)
+        );
+        _require!($cb, description, $lhs => $rhs)
+    }};
+
+    ($cb:expr, $descr:expr, $lhs:expr => $rhs:expr) => {{
+        let rhs = $rhs.to_expr_vec();
+        if rhs.len() == 1 {
+            $cb.require_equal(
+                Box::leak($descr.to_string().into_boxed_str()),
```

Comment:
Not a fan of leaking memory consistently just to have a `'static str`. Isn't there any other way to do this??

Reply:
Perhaps `stringify!($descr)` could work here?

Reply:
Tested yesterday with `$descr` and no error was thrown (see comment above).

Reply:
Unfortunately `stringify!` only works on compile time things, not on dynamic strings. You could try to avoid passing in dynamically created strings, but that severely limits the usefulness of the description in some cases. For example in the EVM circuit we actually do the same leak thing trick for the automatic degree reduction because otherwise the description for those constraints would not be useful. 

But I of course agree that it's far from ideal, and the only good solution is to just modify halo2 to it can accept other types of strings instead of just static ones (I don't know of a good reason why it couldn't just do that).

> Tested yesterday with $descr and no error was thrown (see comment above).

Was still looking into that. The current code I think only uses static strings so that would work, but I do want to allow people to just pass in dynamic strings as well because it's very inconvenient sometimes if it wouldn't accept them.